compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp

```diff
@@ -0,0 +1,324 @@
+      for (uint32_t h = 0; h < nchw_dimension.dim(2).value(); h++)
+      {
+        for (uint32_t w = 0; w < nchw_dimension.dim(3).value(); w++)
+        {
+          uint32_t nchw_indices[4] = {n, c, h, w};
+          uint32_t nhwc_indices[4] = {n, h, w, c};
+          auto data =
+              constant->at<loco::DataType::FLOAT32>(cal_offset(nchw_dimension, nchw_indices));
+          nhwc_const->at<loco::DataType::FLOAT32>(cal_offset(nhwc_dimension, nhwc_indices)) = data;
+        }
+      }
+    }
+  }
+  return nhwc_const;
+}
+
+// We assume ADD with const input is NCHW if,
+// Input shape: (N, C, H, W)
+// Output shape: (N, C, H, W)
+// 1. Const shape is (1, C, 1, 1)
+// 2. Input, Output, Const have the same C.
+bool is_NCHW_with_const(const luci::CircleAdd *node, luci::CircleNode *&pred_node,
+                        luci::CircleConst *&beta)
+{
+  auto x = dynamic_cast<luci::CircleConst *>(node->x());
+  auto y = dynamic_cast<luci::CircleConst *>(node->y());
+
+  if (x != nullptr && y == nullptr)
+  {
+    pred_node = loco::must_cast<luci::CircleNode *>(node->y());
+    beta = x;
+  }
+  else if (x == nullptr && y != nullptr)
+  {
+    pred_node = loco::must_cast<luci::CircleNode *>(node->x());
+    beta = y;
+  }
+  else
+  {
+    // Ignore if ADD does not have a constant input.
+    return false;
+  }
+
+  const auto const_rank = beta->rank();
+  if (const_rank != 4)
+    return false;
+
+  // Check the shape is (1, C, 1, 1)
+  for (uint32_t i = 0; i < const_rank; i++)
+  {
+    if (i == 1)
+      continue;
+
+    if (beta->dim(i).value() != 1)
+      return false;
+  }
+
+  const auto const_cdim = beta->dim(1);
+  const auto input_cdim = pred_node->dim(1);
+  const auto output_cdim = node->dim(1);
+
+  // Check Input, Output, Const have the same channel size
+  if (const_cdim == input_cdim && input_cdim == output_cdim)
+    return true;
+  else
+    return false;
+}
+
+class ConvertNCHWToNHWC final : public luci::CircleNodeMutableVisitor<bool>
+{
+  // Default
+  bool visit(luci::CircleNode *node)
+  {
+    throw std::runtime_error(node->name() + " is an unsupported operator.");
+  }
+
+  bool visit(luci::CircleAdd *node)
+  {
+    luci::CircleNode *pred_node = nullptr;
+    luci::CircleConst *beta = nullptr;
+
+    if (is_NCHW_with_const(node, pred_node, beta))
+    {
+      auto pre_trans = create_pre_transpose(node);
+      pre_trans->a(pred_node);
+
+      auto nhwc_const = create_NHWC_from_NCHW(beta);
+      if (nhwc_const == nullptr)
+        return false;
+
+      node->x(pre_trans);
+      node->y(nhwc_const);
+    }
+    else if (beta == nullptr)
+    {
+      // Both inputs are not constant.
+      // In this case, we cannot distinguish NCHW from NHWC,
+      // so just insert Transpose Ops.
+      auto pre_trans_x = create_pre_transpose(node);
+      pre_trans_x->a(node->x());
+      node->x(pre_trans_x);
+
+      auto pre_trans_y = create_pre_transpose(node);
+      pre_trans_y->a(node->y());
+      node->y(pre_trans_y);
```

Comment:
@jinevening 
You might write this logic because redundant transpose op will be deleted by `RemoveRedundantTranspose` pass.
Say there is this network like this.
```
NHWC - Add(NHWC)  - Conv2D(NHWC) 
```
This network will be like below after this pass
```
NHWC - Transpose - Add(NWCH) - Transpose' - Conv2D(NHWC)
```
This `Transpose` also will be deleted? or it doesn't have to be?

Reply:
> Say there is this network like this.
NHWC - Add(NHWC)  - Conv2D(NHWC) 

The output of onnx-tf does not produce such a pattern. onnx-tf will create a network for Add-Conv2D as below (Transposes are located before/after Conv2D).

`NCHW - Add(NCHW)  - NCHW - Transpose - NHWC - Conv2D(NHWC)  - NHWC - Transpose - NCHW`

This pass will change the network as below (Transposes are added before/after Add).

`NCHW - Transpose - NHWC - Add(NHWC)  - NHWC - Transpose - NCHW - Transpose - NHWC - Conv2D(NHWC)  - NHWC - Transpose - NCHW`

The Transposes between Add and Conv2D can be removed by `RemoveRedundantTransposePass`.

Reply:
@jinevening Thank  you for your explanation. I just assumed a network that has both of NCHW pattern and NHWC one. But this seems overthinking:)